benches/metrics_snapshot.rs

```diff
@@ -0,0 +1,41 @@
+use criterion::{criterion_group, BenchmarkId, Criterion};
+
+fn benchmark(c: &mut Criterion) {
+    let mut group = c.benchmark_group("metrics_snapshot");
+    for &cardinality in [0, 1, 10, 100, 1000, 10000].iter() {
+        group.bench_with_input(
+            BenchmarkId::new("cardinality", cardinality),
+            &cardinality,
+            |b, &cardinality| {
+                let controller = prepare_metrics(cardinality);
+                b.iter(|| {
+                    let iter = vector::metrics::capture_metrics(controller);
+                    assert_cardinality_matches(&iter, cardinality);
```

Comment:
Could we wrap this in `#[cfg(debug_assertions)]` so it isn't included in the benchmarks but is run as a test?

Reply:
If we do it - the iter won't be consumed. I think it's fine, as we're not aiming to measure the `capture_metrics`, but get a sense of what overhead there is in comparison - and since all of the calls include the same logic, and, what's also important, is the logic very simple and scales linearly - I think it should be fine.

The actual assertions in the function could be switched to debug-only, but that can lead to the compiler eliding them. They're also trivial, so I think it's fine too.

Reply:
Ah, yeah, I see. I guess it depends on what we want to benchmark here. If consuming the iterator is important, we could probably just return a `.collect()` to `b.iter()` here.

Reply:
> just return a `.collect()` to `b.iter()` here

That's what's happening under the hood, pretty much. The overhead of the `assert_cardinality_matches` beyond consuming the iteration should be negligible.

Reply:
Gotcha :+1: Could we just add a comment to that effect then? I think all other benchmark assertions are currently only done when `debug_assertions` is enabled (e.g. in `cargo test`).

Reply:
Sure!
Btw, I just realized we don't even consume the iteratir there, but merely check it's length (which is also fine).

Reply:
Added a comment at 47c54abce63ba4dc381e257a5ae762ba12f73616